pkg/pipeline/transform/kubernetes/cni/ovn_kubernetes.go

```diff
@@ -43,27 +43,49 @@ func AddOvnIPs(ips []string, node *v1.Node) []string {
 	return ips
 }
 
+func unmarshalAnnotation(annot []byte) (string, error) {
+	// Depending on OVN (OCP) version, the annotation might be JSON-encoded as a string, or an array of strings
+	// Try first as string (legacy)
+	var subnetsAsString map[string]string
+	err := json.Unmarshal(annot, &subnetsAsString)
+	if err == nil {
+		if subnet, ok := subnetsAsString["default"]; ok {
+			return subnet, nil
+		}
+		return "", fmt.Errorf("unexpected content for annotation %s: %s", ovnSubnetAnnotation, annot)
+	}
+
+	var subnetsAsArray map[string][]string
+	err = json.Unmarshal(annot, &subnetsAsArray)
+	if err == nil {
+		if subnets, ok := subnetsAsArray["default"]; ok {
+			if len(subnets) > 0 {
+				return subnets[0], nil
+			}
+		}
+		return "", fmt.Errorf("unexpected content for annotation %s: %s", ovnSubnetAnnotation, annot)
+	}
+
+	return "", fmt.Errorf("cannot read annotation %s: %v", ovnSubnetAnnotation, err)
+}
+
 func findOvnMp0IP(annotations map[string]string) (string, error) {
 	if subnetsJSON, ok := annotations[ovnSubnetAnnotation]; ok {
-		var subnets map[string]string
-		err := json.Unmarshal([]byte(subnetsJSON), &subnets)
+		subnet, err := unmarshalAnnotation([]byte(subnetsJSON))
 		if err != nil {
-			return "", fmt.Errorf("cannot read annotation %s: %v", ovnSubnetAnnotation, err)
+			return "", err
 		}
-		if subnet, ok := subnets["default"]; ok {
-			// From subnet like 10.128.0.0/23, we want to index IP 10.128.0.2
-			ip0, _, err := net.ParseCIDR(subnet)
-			if err != nil {
-				return "", err
-			}
-			ip4 := ip0.To4()
-			if ip4 == nil {
-				// TODO: what's the rule with ipv6?
-				return "", nil
-			}
-			return fmt.Sprintf("%d.%d.%d.2", ip4[0], ip4[1], ip4[2]), nil
+		// From subnet like 10.128.0.0/23, we want to index IP 10.128.0.2
+		ip0, _, err := net.ParseCIDR(subnet)
+		if err != nil {
+			return "", err
+		}
+		ip4 := ip0.To4()
+		if ip4 == nil {
+			// TODO: what's the rule with ipv6?
+			return "", nil
 		}
-		return "", fmt.Errorf("unexpected content for annotation %s: %s", ovnSubnetAnnotation, subnetsJSON)
+		return fmt.Sprintf("%d.%d.%d.2", ip4[0], ip4[1], ip4[2]), nil
```

Comment:
what is `.2` ? maybe u can use `fmt.Sprintf("%s",net.IPv4(ip4))`

Reply:
The IP here is the first one of the range provided by the subnet that is read above; e.g. we read the subnet `10.128.0.0/23` ; then it's a tacit rule with ovn-k that the `.2` ip from that subnet is used for the mp0 interface, which is the one we want to track here. So we need somehow to convert `10.128.0.0/23` into `10.128.0.2`.

Reply:
That will work only if you can assume the suffix is never greater than 24.  Otherwise, it should add 2 or | 2.

Reply:
From past discussions, this is an assumption that we can make - although there's no guarantee that it won't change.